internals/overlord/logstate/gatherer.go

```diff
@@ -133,17 +133,23 @@ func fillDefaultOptions(options *logGathererOptions) *logGathererOptions {
 // PlanChanged is called by the LogManager when the plan is changed, if this
 // gatherer's target exists in the new plan.
 func (g *logGatherer) PlanChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	// TODO: we should flush the client here because the labels may have changed.
+	// But this is done in a separate goroutine to the main loop - is that going
+	// to be a problem?
```

Comment:
I believe it's a problem even as is, because `lokiClient.SetLabels` updates the map, and `lokiClient.Flush` accesses it, which is a data race.

I think the simplest option would be to define the `SetLabels` method (and only that method) as being safe to call concurrently with `Write`/`Flush`. Then have a single mutex in the Loki client that locked when `Flush` uses the labels map and when `SetLabels` updates it.

Reply:
I don't think that will solve the problem, as here we've still got a `Flush` being triggered from `PlanChanged`, potentially concurrent to flushes from the main loop (so concurrent reading/writing of the buffer/entries slice).

Not to mention that I'd rather keep the mutex outside the client, in the gatherer. This avoids us having to make all our future client implementations concurrency-safe. Generally, we should put shared/common logic in the gatherer to make clients as small (less code) as possible.

Reply:
Sounds reasonable. Give it a shot and let's see what that looks like!